doc/reference/trd-dynamic-process-loading.md

```diff
@@ -0,0 +1,340 @@
+└────────────────────────┘  └────────────────────────┘                         
+```
+
+The layers provide differing levels of trust for each component.
+
+- **Userspace Application**: The userspace application is generally untrusted.
+  It is responsible for obtaining the process binary. Generally, this
+  application is expected to work correctly, but it may not be able to verify a
+  process binary it receives.
+- **AppLoader Capsule**: The `AppLoader` capsule has normal capsule trust and
+  provides a system call interface for dynamic process loading. Because it is a
+  conventional capsule, we do not trust it to enforce the Tock Threat Model
+  requirements.
+- **Kernel Capsules**: The kernel capsules are the minimal functionality that
+  _must_ be trusted to ensure the Tock security guarantees are met. This
+  includes verifying process credentials and ensuring the process storage format
+  remains valid.
+
+
+3 System Call API
+=================================
+
+The `0x10001` system call interface provides four operations:
+
+1. `setup(process_binary_size_bytes: usize)`: This initiates the process of
+   loading a new process binary. The `AppLoader` capsule will attempt to allocate resources
+   to be able to store and load a process of the specified size.
+   Each process may only setup one process binary to load at a time.
+   Success or
+   failure is indicated via an upcall.
+2. `write(process_binary: &[u8], length_bytes: usize, offset_bytes: usize)`:
+   This stores a portion of the process binary.
+   `setup()` must have completed successfully before `write()` can be called.
+   Write is expected to be called
+   multiple times to store the entire process binary and there is no assumption
+   about the order the process binary is written in. Success or failure is
+   indicated via an upcall.
+3. `load()`: This indicates the entire process binary has been written and the
+   new process binary should be loaded into a process. Success or failure is
+   indicated via an upcall.
+4. `abort()`: This operation cancels the setup/write operation and frees 
+   allocated resources so that they are available for a different process. 
+   Success or failure is indicated via an upcall.
+
+These operations are implemented using conventional allow, command, and
+subscribe system calls.
+
+
+4 Kernel Capsule Traits
+=================================
+
+The trusted capsules use the following traits to provide functionality to the
+app loading capsule.
+
+4.1 Dynamic Process Binary Storing
+---------------------------------
+
+The kernel provides an interface for storing a process binary loaded at runtime.
+The kernel must expose this interface to ensure that the implementation
+correctly upholds the Tock Threat Model guarantees. Specifically, storing a new
+process binary must not cause existing process binaries to be "lost" or not
+loaded on future reboots of the board.
```

Comment:
This only covers the `integrity` half of the threat model, I think this also either (a) needs to elucidate thoughts on `confidentiality`, or more simply (b), just refer to the Threat Model, i.e.:
```suggestion
correctly upholds the Tock Threat Model guarantees.
```